src/library/scala/collection/immutable/Vector.scala

```diff
@@ -1559,6 +1566,108 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
     this
   }
 
+  //TODO Make public; this method is only private for binary compatibility
+  private[collection] def alignTo(before: Int, bigVector: Vector[A]): this.type = {
+    if (len1 != 0 || lenRest != 0)
+      throw new UnsupportedOperationException("A non-empty VectorBuilder cannot be aligned retrospectively. Please call .reset() or use a new VectorBuilder.")
+    val (prefixLength, maxPrefixLength) = bigVector match {
+      case Vector0 => (0, 1)
+      case v1: Vector1[_] => (0, 1)
+      case v2: Vector2[_] => (v2.len1, WIDTH)
+      case v3: Vector3[_] => (v3.len12, WIDTH2)
+      case v4: Vector4[_] => (v4.len123, WIDTH3)
+      case v5: Vector5[_] => (v5.len1234, WIDTH4)
+      case v6: Vector6[_] => (v6.len12345, WIDTH5)
+    }
+    if (maxPrefixLength == 1) return this // does not really make sense to align for <= 32 element-vector
+    val overallPrefixLength = (before + prefixLength) % maxPrefixLength
+    offset = (maxPrefixLength - overallPrefixLength) % maxPrefixLength
+    // pretend there are already `offset` elements added
+    advanceN(offset & ~MASK)
+    len1 = offset & MASK
+    prefixIsRightAligned = true
+    this
+  }
+
+  /**
+   * Removes `offset` leading `null`s in the prefix.
+   * This is needed after calling `alignTo` and subsequent additions,
+   * directly before the result is used for creating a new Vector.
+   *
+   * example:
+   *     a2 = Array(null, ..., null, Array(null, .., null, 0, 1, .., x), Array(x+1, .., x+32), ...)
+   * becomes
+   *     a2 = Array(Array(0, 1, .., x), Array(x+1, .., x+32), ...)
+   */
+  private[this] def leftAlignPrefix(): Unit = {
+    var a: Array[AnyRef] = null // the array we modify
+    var aParent: Array[AnyRef] = null // a's parent, so aParent(0) == a
+    if (depth >= 6) {
+      a = a6.asInstanceOf[Array[AnyRef]]
+      val i = offset >>> BITS5
+      if (i > 0) {
+        a = copyOfRange(a, i, LASTWIDTH)
+        a6 = a.asInstanceOf[Arr6]
+      }
+      aParent = a
+      a = a(0).asInstanceOf[Array[AnyRef]]
+    }
+    if (depth >= 5) {
+      if (a == null) a = a5.asInstanceOf[Array[AnyRef]]
+      val i = (offset >>> BITS4) & MASK
+      if (i > 0) {
+        a = copyOfRange(a, i, WIDTH)
+        if (depth == 5) a5 = a.asInstanceOf[Arr5]
```

Comment:
I'm doing some "poor man's test coverage" checks by replacing expressions with `???`

The `true` case of this `if` is not covered by the unit tests, nor is the analagous line for other depths below. What data would be need to get here?

Reply:
Here's the spots I found that might need more variety in the test case data:

```diff
diff --git a/src/library/scala/collection/immutable/Vector.scala b/src/library/scala/collection/immutable/Vector.scala
index 7dd3334e7d..db88f4158b 100644
--- a/src/library/scala/collection/immutable/Vector.scala
+++ b/src/library/scala/collection/immutable/Vector.scala
@@ -1608,6 +1608,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
       if (i > 0) {
         a = copyOfRange(a, i, LASTWIDTH)
         a6 = a.asInstanceOf[Arr6]
+        ???
       }
       aParent = a
       a = a(0).asInstanceOf[Array[AnyRef]]
@@ -1617,7 +1618,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
       val i = (offset >>> BITS4) & MASK
       if (i > 0) {
         a = copyOfRange(a, i, WIDTH)
-        if (depth == 5) a5 = a.asInstanceOf[Arr5]
+        if (depth == 5) ???
         else aParent(0) = a
       }
       aParent = a
@@ -1628,7 +1629,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
       val i = (offset >>> BITS3) & MASK
       if (i > 0) {
         a = copyOfRange(a, i, WIDTH)
-        if (depth == 4) a4 = a.asInstanceOf[Arr4]
+        if (depth == 4) ???
         else aParent(0) = a
       }
       aParent = a
@@ -1639,7 +1640,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
       val i = (offset >>> BITS2) & MASK
       if (i > 0) {
         a = copyOfRange(a, i, WIDTH)
-        if (depth == 3) a3 = a.asInstanceOf[Arr3]
+        if (depth == 3) ???
         else aParent(0) = a
       }
       aParent = a
@@ -1650,7 +1651,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
       val i = (offset >>> BITS) & MASK
       if (i > 0) {
         a = copyOfRange(a, i, WIDTH)
-        if (depth == 2) a2 = a.asInstanceOf[Arr2]
+        if (depth == 2) ???
         else aParent(0) = a
       }
       aParent = a
@@ -1661,7 +1662,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
       val i = offset & MASK
       if (i > 0) {
         a = copyOfRange(a, i, WIDTH)
-        if (depth == 1) a1 = a.asInstanceOf[Arr1]
+        if (depth == 1) ???
         else aParent(0) = a
       }
     }
@@ -1757,9 +1758,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
         }
       case 6 => // note width is now LASTWIDTH
         if (lenRest % WIDTH5 != 0) {
-          // lenRest is not multiple of WIDTH5, so this slice does not align, need to try lower dimensions
-          slice.foreach(e => addArrN(e.asInstanceOf[Array[AnyRef]], 5))
-          return
+          ???
         }
         val copy1   = mmin((BITS * 6 + 1 - lenRest) >>> BITS5, sl)
         val copy2   = sl - copy1
@@ -1767,8 +1766,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
         System.arraycopy(slice, 0, a6, destPos, copy1)
         advanceN(WIDTH5 * copy1)
         if (copy2 > 0) {
-          System.arraycopy(slice, copy1, a6, 0, copy2)
-          advanceN(WIDTH5 * copy2)
+          ???
         }
     }
   }
```

Reply:
> What data would be need to get here?

Should require be something like:
* The builder is filled up to `a5` (but not `a6`)
* `offset` is so large that `a5` starts with a `null` entry

Off the top of my head, I don't think this should occur in normal operation, as in such a case, [the modulo in `alignTo`](https://github.com/scala/scala/blob/e4b872afd9d197daf9e73dca0fbd289e2cf49d71/src/library/scala/collection/immutable/Vector.scala#L1584) should have prevented that. The only case that comes to my mind is if somebody called `alignTo` (when it's eventually public) with a different vector than what they later add to the builder. That is not likely to make sense but isn't really forbidden either. Maybe something like

```scala
  val v6 = ... // Vector6
  val vb = new VectorBuilder()
  vb.alignTo(23, v6)
  ab.addOne(..) ...// add 23 values
  val v5 = dropSomethingFromTheRight(v6) // drops so much that we now have a Vector5, but the offset is still set
  vb.addAll(v5)
  vb.result()
```

Anyway, good catch! I'm gonna try to find test cases for all the uncovered branches.

Reply:
As mentioned, the branches were supposed to handle "misleading" arguments to `alignTo`. Problem was that `result()` cannot handle the `null` entries this created. Thus, if necessary, I now move the entries to the left and lower `offset` (and `lenRest`) accordingly. In some cases this may even reduce the `depth`.
```scala
        shrinkOffsetIfToLarge(WIDTH4)
        if ((lenRest >>> BITS4) == 0) depth = 4
```
Also, I now use `System.arraycopy` to keep `a` at it's current size. This is necessary to allow reusing the Builder after a call to `result()`.
```scala
        if (i > 0) System.arraycopy(a, i, a, 0, WIDTH - i)
```

This should explain most of the diff in `leftAlignPrefix()`.

---

The second of the two other branches you found, was actually useless, I removed it.

---

Every branch you mentioned should be covered by tests now.